File: GO/leetcode/leetcodeE/783_MinimumDistanceBetweenBSTNodes.go
package leetcodeE

import (
	"math"
)

type Stones [2]int

// res := []int{math.MinInt32, math.MaxInt32}

func (s Stones) search(node *TreeNode) {

	if node == nil {
		return
	}

	s.search(node.Left)

	if node.Val-s[0] < s[1] {
		s[1] = node.Val - s[0]
	}

	s.search(node.Right)
}

func minDiffInBST1(root *TreeNode) int {

	s := Stones{math.MinInt32, math.MaxInt32}
	s.search(root)
	return s[1]
}

var res []int

func searchTree(node *TreeNode) {

	if node == nil {
		return
	}

	searchTree(node.Left)

	if node.Val-res[0] < res[1] {
		res[1] = node.Val - res[0]
	}
	res[0] = node.Val

	searchTree(node.Right)
}

func minDiffInBST2(root *TreeNode) int {

	res = []int{math.MinInt32, math.MaxInt32}
	searchTree(root)
	return res[1]
}

File: PythonLeetcode/LeetCodeE/783_MinimumDistanceBetweenBSTNodes.py
# 783. Minimum Distance Between BST Nodes

# Given a Binary Search Tree (BST) with the root node root, return the minimum difference 
# between the values of any two different nodes in the tree.

# Example :

# Input: root = [4,2,6,1,3,null,null]
# Output: 1
# Explanation:
# Note that root is a TreeNode object, not an array.

# The given tree [4,2,6,1,3,null,null] is represented by the following diagram:

#          4
#        /   \
#      2      6
#     / \    
#    1   3  

# while the minimum difference in this tree is 1, it occurs between node 1 and node 2, also between node 3 and node 2.


class MinDiffInBST:

    def doit(self, root: TreeNode) -> int:
        
        def search(node):
            
            if not node.left and not node.right:
                return [node.val, node.val, float('inf')]
            
            res = [node.val, node.val, float('inf')]
            
            if node.left:
                c = search(node.left)
                res[0] = c[0]
                res[2] = min(c[2], node.val - c[1])
            if node.right:
                c = search(node.right)
                res[1] = c[1]
                res[2] = min(c[2], c[0] - node.val, res[2])
            
            return res
    
        return search(root)[2]

    def doit1(self, root: TreeNode) -> int:

        buff = [float('-inf'), float('inf')]

        def search(node):

            if not node:
                return

            search(node.left)

            buff[1] = min(node.val - buff[0], buff[1])
            buff[0] = node.val
            
            search(node.right)

        search(root)
        return buff[1]